R5 update

borisfom opened the pull request:
Our latest R5 developments + master updates

soumith commented:
Boris, if you notice there are 85 commits in this PR. This is because you tried to Merge 'master' into R5. Instead, you should rebase your changes on top of R5 so that this issue does not occur. 
Can you do that, or do you want me to do that?

borisfom commented:
Probably, the best way would be if you rebase R5 to master, and I would resubmit request (probably from a new branch - R5 in my repo is release branch for torch-nv so I won't mess with its history ?

nicholas-leonard commented:
@borisfom Does the new `nn.RNN` also implement LSTM functionality? Or that will be included in a separate PR?

borisfom commented:
Yes RNN also implements LSTM. Here's a brief description from the authors: 
"By  default the module uses RNN ReLU, but the user can configure it to use LSTM or GRU. Ideally there would be separate Torch modules pre-configured for each RNN mode that would allow for the user to access mode-specific information about the parameters (using cudnnGetRNNLinLayer{Matrix,Bias}Params), but this is outside the scope of a first implementation.
 
The module has been sanity tested by comparing the output with that of cuDNN samples, but there are no unit tests. This should be sufficient for now since cuDNN is tested. In the long term, a Torch reference RNN implementation/unit test should also be made."

borisfom commented:
Soumith: I have rebased your R5 on top of master here: https://github.com/borisfom/cudnn.torch/commits/R5_up.  (There is 43 commits). Should I simply push it back to R5 or use PR for that, or leave it alone (see below) ? 
Then I have rebased my R5 on top of that (18 extra commits). 
However, this PR now became 43 commits altogether - should we just merge it ?

soumith commented:
give me time till monday / tuesday to review it. have been busy with GTC.

szagoruyko commented on Pointwise.lua:
```diff
@@ -69,5 +88,7 @@ end
 function Pointwise:clearState()
    self:clearDesc()
+   self._gradOutput = nil
+   nn.utils.clear(self, 'activDesc')
```
`activDesc` is niled in `clearDesc` already, should be `_gradOutput` there instead so that `nn.utils.clear` can call `set` on it

borisfom replied:
Do you think _gradOutput requires clear() ? In other places it's just being set to nil.

szagoruyko replied:
yes it's better to use clear() to allow buffer sharing, see https://github.com/torch/nn/pull/526

szagoruyko commented on README.md:
```diff
@@ -86,8 +86,7 @@ nn.Sequential {
 For version CuDNN R1, checkout the branch **R1**
 For version CuDNN R2, checkout the branch **R2**
 For version CuDNN R3, checkout the branch **R3**
+For version CuDNN R3, checkout the branch **master**
```
CuDNN R4?

szagoruyko commented on SpatialFullConvolution.lua:
```diff
@@ -8,43 +8,20 @@ autotunerCache[1] = {} -- forward
-    self.reset = nil
-    self.iSize = torch.LongStorage(4):fill(0)
-end
-
```
this removes groups support ! that means that we will not be able to run any alexnet networks in torch.

szagoruyko replied:
ignore this, didn't notice this is FullConvolution, phew

szagoruyko commented:
I did a pass, looks good. the change in temporary buffers is in https://github.com/soumith/cudnn.torch/pull/162

SeanNaren commented:
Just a heads up, there are a few changes that can be done to the RNN module to make it easier to create BRNNs GRUs etc. If you change the local functions/global functions to passing self in (via colon or explicitly) you can inherit a class from it. i.e change local/functions such as this:
```
local function createDropoutDescriptors(count)
   return createDescriptors(count,
                            'cudnnDropoutDescriptor_t[?]',
                            'cudnnCreateDropoutDescriptor',
                            'cudnnDestroyDropoutDescriptor')
end
```
To this:
```
function RNN:createDropoutDescriptors(count)
    return self:createDescriptors(count,
        'cudnnDropoutDescriptor_t[?]',
        'cudnnCreateDropoutDescriptor',
        'cudnnDestroyDropoutDescriptor')
end
```

Also to make it easier to support BRNNs exposing the resizing of outputs (to add *2 to the hidden size) may be a good idea. To make it clearer I've already implemented it in my fork in the RNN class [here](https://github.com/borisfom/cudnn.torch/blob/R5/RNN.lua).

I've also used the above changes to create a working BLSTM [here](https://github.com/SeanNaren/cudnn.torch/blob/R5/BLSTM.lua). This may be more useful in a future PR but hopefully this is somewhat useful and great work on the PR so far!

borisfom commented:
Well, looks like my rebases did not quite make it less commits - please take a stab on it as you see fit!

szagoruyko commented on ffi.lua:
```diff
@@ -660,16 +723,15 @@ cudnnStatus_t  cudnnSoftmaxBackward(
-    CUDNN_POOLING_AVERAGE_COUNT_EXCLUDE_PADDING = 2, /*  count for average does not include padded values */
-    CUDNN_POOLING_AVERAGE = CUDNN_POOLING_AVERAGE_COUNT_INCLUDE_PADDING /*  for backward compatibility */
+    CUDNN_POOLING_AVERAGE_COUNT_INCLUDE_PADDING = 1, /* count for average includes padded values*/
+    CUDNN_POOLING_AVERAGE_COUNT_EXCLUDE_PADDING = 2 /* count for average does not include padded values*/
```
breaks backward compatibility with saved models

borisfom replied:
That's what CUDNN5 has. We certainly can add backwards compatible CUDNN_POOLING_AVERAGE.

szagoruyko commented:
I rebased @borisfom changes in a single commit here https://github.com/szagoruyko/cudnn.torch/tree/R5n without RNN class, I suggest we make it R5 and do a separate PR for RNN to review thoroughly and add tests (I didn't find any)

soumith commented:
@szagoruyko that looks clean. let's do that.

borisfom commented:
@szagoruyko: sounds like a plan! Also, I plan to merge a PR that has RNN tests and adds a few other classes based on RNN: https://github.com/borisfom/cudnn.torch/pull/3

szagoruyko commented:
@soumith done https://github.com/soumith/cudnn.torch/tree/R5
@borisfom @SeanNaren great stuff! please do a PR to the new R5 branch
